hc_api/src/lib.rs

```diff
@@ -0,0 +1,259 @@
+/*!
+hc_api provides a library for container applications to instantiate and run holochain applications.
+
+# Examples
+
+``` rust
+extern crate hc_core;
+extern crate hc_api;
+extern crate hc_dna;
+extern crate hc_agent;
+
+use hc_api::*;
+use hc_dna::Dna;
+use hc_agent::Agent;
+use std::sync::{Arc, Mutex};
+use hc_core::context::Context;
+use hc_core::logger::SimpleLogger;
+
+// instantiate a new app
+
+// need to get to something like this:
+//let dna = hc_dna::from_package_file("mydna.hcpkg");
+
+// but for now:
+let dna = Dna::new();
+let agent = Agent::from_string("bob");
+let context = Context {
+    agent: agent,
+    logger: Arc::new(Mutex::new(SimpleLogger {})),
+};
+let mut hc = Holochain::new(dna,Arc::new(context)).unwrap();
+
+// start up the app
+hc.start().expect("couldn't start the app");
+
+// call a function in the app
+hc.call("some_fn");
+
+// get the state
+{
+    let state = hc.state();
+
+    // do some other stuff with the state here
+    // ...
+}
+
+// stop the app
+hc.stop().expect("couldn't stop the app");
+
+```
+*/
+
+extern crate hc_agent;
+extern crate hc_core;
+extern crate hc_dna;
+
+use hc_core::context::Context;
+use hc_dna::Dna;
+use std::sync::Arc;
+
+/// contains a Holochain application instance
+#[derive(Clone)]
+pub struct Holochain {
+    instance: hc_core::instance::Instance,
+    context: Arc<hc_core::context::Context>,
+    active: bool,
+}
+
+use hc_core::error::HolochainError;
+use hc_core::nucleus::fncall;
+use hc_core::nucleus::Action::*;
+use hc_core::state::Action::*;
+use hc_core::state::State;
+
+impl Holochain {
+    /// create a new Holochain instance
+    pub fn new(dna: Dna, context: Arc<Context>) -> Result<Self, HolochainError> {
+        let mut instance = hc_core::instance::Instance::new();
+        let action = Nucleus(InitApplication(dna.clone()));
+        instance.dispatch(action);
+        instance.consume_next_action()?;
```

Comment:
So this will change in my branch, but I think we should finish and merge this one first and and then I will merge develop into my branch and adjust..